File: 3958-MinimumRemovalsToBalanceArray/3958-MinimumRemovalsToBalanceArray.java
// Last updated: 6/10/2026, 7:11:03 PM
class Solution {
    public int minRemoval(int[] nums, int k) {
        Arrays.sort(nums);
        int n = nums.length;
        // int ans = Integer.MAX_VALUE;
        int left = 0, mx = 0;
        for(int r=0 ; r<n ; r++) {
            while(nums[r] > 1L * nums[left] * k) left++;
            mx = Math.max(mx , r - left + 1);
        }
        return n - mx;
    }
}

File: 3958-MinimumRemovalsToBalanceArray/3958-MinimumRemovalsToBalanceArray.cpp
// Last updated: 6/10/2026, 7:11:09 PM
class Solution {
public:
    int minRemoval(vector<int>& nums, int k) {
        int ans = INT_MAX;
        sort(nums.begin() , nums.end());
        for(int i=0 ; i<nums.size() ; i++) {
            auto it = upper_bound(nums.begin() , nums.end() , 1LL * k * nums[i]);
            it = prev(it);
            // if(it == nums.end() || *it > 1LL * k * nums[i]) it = prev(it);
            int right = nums.size() - (it - nums.begin()) - 1;
            ans = min(ans , i + right);
        }
        return ans;
    }
};